Exception handler for Telemetry SDK

alfwatt opened the pull request:
- Stubs documentation comments for the existing Public API
- Marks some Events constructors as to be deprecated, pending replacements
- Cleanup of MMETestHost
- Update name headers, remove empty class

Commit:
[iOS] Exception handler for Telemetry SDK mapbox/mobile-telemetry#287

- Better imports in the .h, better import order in the .m
- Group the deprecated event constructors, document the rest
- Add scripts to project
- MapboxMobileMetrics -> MapboxMobileEvents
- Add MMEErrorDomain and associated values to MMEConstants
- MMEEvent marks a number of deprecated methods for later warning flags
- MMEEvent routes all event construction through to the designated initializer which now reports errors
- MMEEventsManager adds Error & Exception Reporting
- Some tests now fail because MMEEvent is immutable, will update
- Move deviceOrientation and contentSizeScale into UIKit+MMEMobileEvents
- Add dateWithDate: to MMEDate
- Event equality uses the storage properties for efficiency, does not separately consider name
- Fix some of the broken tests
- Don't break encapsulation in MMEvent or MMEDate
- Add MMEDateFakes to allow for sloppy matching of dates in tests
- Group Categories in a Folder
- Fix Static Analyzer warnings in APIClient
- Add MME_DEPRECATED_GOTO
- Xcode 10.2 updates and English -> en conversion
- Add MMECrashTests.mm  for basic crash event testing
- Add `+ commentEventData` to MMECommonEventData
- Fix tests which were broken on device
- Turn on code coverage for testing in the Xcode project
- Add NS_ASSUME_NONNULL blocks on MMEEvent, update tests to reflect
- Fix nested test cases for encode decide of dates and events
- Implement all the @try{} blocks for the MMEventsManager Exception Free API

rclee commented on MapboxMobileEvents/MMEEvent.h:
```diff
@@ -17,26 +18,96 @@
 + (instancetype)carplayEventWithName:(NSString *)name attributes:(NSDictionary *)attributes;
-+ (instancetype)eventWithDateString:(NSString *)dateString name:(NSString *)name attributes:(NSDictionary *)attributes;
 
+#pragma mark - Depricated (Date Strings)
```
typo: Depricated  -> Deprecated

rclee commented on MapboxMobileEvents/MMEEventsManager.h:
```diff
@@ -1,55 +1,134 @@
 
+/*! @brief displayLogFileFromDate:...
+    @param logDate earliest date to display log information for
+    @note TO BE DEPRICATED
```
typo: DEPRICATED -> DEPRECATED

alfwatt commented on MapboxMobileEvents.xcodeproj/xcshareddata/xcschemes/MMETestHost.xcscheme:
```diff
@@ -0,0 +1,91 @@
+      </AdditionalOptions>
+   </LaunchAction>
+   <ProfileAction
+      buildConfiguration = "Release"
```
It should be possible to run the tests in a Release build configuration, so I think this is OK.

rclee commented on MapboxMobileEvents/MMEEvent.h:
```diff
@@ -86,28 +68,90 @@
+
+#pragma mark - Deprecated (MMECommonEventData)
+
+/*! @brief deprecated in MabboxMobileEvents 1.0.0 or later
```
typo: MabboxMobileEvents -> MapboxMobileEvents

rclee replied:
Still have Mabbox going on here

rclee commented on MapboxMobileEvents/MMEEvent.h:
```diff
@@ -86,28 +68,90 @@
-    @param name name
-    @param attributes attrs
-    @return event
+/*! @brief deprecated in MabboxMobileEvents 1.0.0 or later
```
and here: MabboxMobileEvents -> MapboxMobileEvents

rclee replied:
More below this one as well. Copy pasta 🍝

rclee commented on MapboxMobileEvents/MMETypes.h:
```diff
@@ -22,3 +22,21 @@
 typedef NS_DICTIONARY_OF(NSString *, id) MMEMapboxEventAttributes;
 typedef NS_MUTABLE_DICTIONARY_OF(NSString *, id) MMEMutableMapboxEventAttributes;
+
+#ifdef MBE_1_0_0_RELEASE
```
nit: I think it might make more sense to have MBE ( assuming `MBE_1_0_0_RELEASE` stands for Mapbox Events?) as MME (Mapbox Mobile Events).

rclee commented on MapboxMobileEvents/MMEEvent.m:
```diff
@@ -3,206 +3,192 @@
 
 + (instancetype)telemetryMetricsEventWithDateString:(NSString *)dateString attributes:(NSDictionary *)attributes {
-    MMEEvent *telemetryMetrics = [[MMEEvent alloc] init];
-    telemetryMetrics.name = MMEEventTypeTelemetryMetrics;
```
I think the main use case for the event `name` being added to the events in this way was for debug logging. It allows for some easy access to the event name without having to go through `attributes`.

rclee reviewed (commented):
Minor things :)

rclee commented on MapboxMobileEvents/MMEEvent.m:
```diff
@@ -3,305 +3,336 @@
+        self = [super init];
+    }
+    else if ([NSJSONSerialization isValidJSONObject:eventAttributes]) {
+        if (![eventAttributes.allKeys containsObject:MMEEventKeyEvent]) { // requried
```
typo: requried -> required (hope these aren't too annoying! Hard for me to pass by one if I see it 😅)

rclee commented on MapboxMobileEventsTests/MMECrashTests.mm:
```diff
@@ -0,0 +1,37 @@
+
+describe(@"MMEEventsMangaer", ^{
+
+    context(@"- reportError:", ^{
```
These tests should follow the same [structuring](https://github.com/cedarbdd/cedar/wiki/Writing-specs) as our other specs. We're missing `beforeEach` (if needed) and `it` for each assert (`testEvent should_not be_nil;` in this example)

Can install some [handy snippets](https://github.com/cedarbdd/cedar/wiki/Getting-started#code-snippets) here if you haven't already

rclee commented on MapboxMobileEventsTests/MMEDateFakes.m:
```diff
@@ -0,0 +1,101 @@
+}
+
+- (NSComparisonResult) compare:(NSDate *)date {
+    NSComparisonResult comprable = NSOrderedSame;
```
nit typo: comprable (except maybe by brits?) -> comparable

alfwatt replied:
I worship the old English

rclee reviewed (commented):
Looking good. Loving the improvements!

julianrex commented on MapboxMobileEvents/Categories/UIKit+MMEMobileEvents.m:
```diff
@@ -0,0 +1,84 @@
+
+    NSString *sc = self.preferredContentSizeCategory;
+
+    if ([sc isEqualToString:UIContentSizeCategoryExtraSmall]) {
```
This might be clearer as an `NSDictionary` and lookup.

alfwatt replied:
a dictionary would use a little more memory and a little less runtime, going to leave it for now

julianrex commented on MapboxMobileEvents/MMECommonEventData.m:
```diff
@@ -26,12 +26,43 @@ + (NSString *)sysInfoByName:(char *)typeSpecifier {
+    NSMutableDictionary *eventDictionary = NSMutableDictionary.new;
+    MMECommonEventData *eventData = MMECommonEventData.new;
+
+    if (eventData.model) {
```
IIRC the `if` checks here are redundant, since `eventDictionary` is a new empty dict - so you wouldn't be removing anything.

alfwatt replied:
`eventData` is an instance of `MMECommonEventData` which sets up those properties in `init` just below.

julianrex commented on MapboxMobileEvents/MMEEvent.m:
```diff
@@ -3,305 +3,336 @@
+    NSError *eventError = nil;
+    MMEEvent *newEvent = [MMEEvent eventWithAttributes:attributes error:&eventError];
+    if (eventError != nil) {
+        [MMEEventsManager.sharedManager reportError:eventError];
```
It doesn't feel right that the `MMEEvent` knows about (and calls) `MMEEventsManager`. Perhaps this should be a method on `MMEEventsManager`?

Also, with `sharedManager` is the intention that consuming SDKs should not instantiate their own copy?

julianrex replied:
Given that there's the `eventWithAttributes:error:` method below, this one feels redundant.

alfwatt replied:
Ok, the `sharedManager` issue is a problem, we rely on it for telemetry metrics collection and might be missing a lot of those due to how it's used in the maps SDK right now. The example in the Readme is wrong, the correct way to initialize the class (since moving over to relying on the singleton internally) is:

    [MMEEventsManager.sharedManager initializeWithAccessToken:userAgentBase:hostSDKVersion:]

In `MGLMapboxEvents.m:48` a new instance is created when the shared instance should be used.

alfwatt replied:
I've updated the readme and will make the change in the MapsSDK when I open the integration ticket.

julianrex commented on MapboxMobileEvents/MMEEvent.m:
```diff
@@ -3,305 +3,336 @@
+    crashAttributes[MMEEventKeyStackTraceHash] = [NSString stringWithFormat:@"%@/%ld %@ %@", eventsError.domain, (long)eventsError.code,
+        (eventsError.localizedDescription ?: @"") ,
+        (eventsError.localizedFailureReason ?: MMEEventKeyErrorNoReason)];
+    crashAttributes[MMEEventKeyAppID] = (NSBundle.mainBundle.bundleIdentifier ?: @"unknown");
```
I wonder if `bundleIdentifier` etc should be passed to the Telemetry SDK

julianrex commented:
@alfwatt just added some comments I had pending; I see there have been some changes since I first wrote those, so apologies if any are out of date 😢

alfwatt commented on MapboxMobileEvents/MMEDate.m:
```diff
@@ -121,6 +125,24 @@ - (NSString *)description {
     return [NSString stringWithFormat:@"<%@ sinceReference=%f, offsetFromServer=%f>", NSStringFromClass(self.class), _sinceReferenceDate, _offsetFromServer];
 }
 
+- (BOOL) isEqual:(id)object {
```
That case is taken care of internally, there's a small performance penalty for doing the type check, maybe if we find it's a performance issue at some point?

alfwatt commented on MapboxMobileEvents/MMEEvent.m:
```diff
@@ -3,305 +3,336 @@
++ (instancetype)crashEventReporting:(NSError *)eventsError error:(NSError **)createError {
+    // start with common event data
+    NSMutableDictionary *crashAttributes = MMECommonEventData.commonEventData;
+    crashAttributes[MMEEventKeyEvent] = MMEEventMobileCrash;
```
Correct, the exceptions are wrapped up into `NSError` objects, and we generate a report from that, so that networking, filesystem or encode/decode errors can be collected as well.

`PLCrashReporter` for e.g. could be added to catch crashes, then this method used to send back a report on app launch if a crash file is left behind.

alfwatt commented:
@julianrex that's a huge help, working through these issue now

Commit:
Add docs generation makefile rules, updates dates in LICENSE.md

Commit:
MMEAPIClient should invalidate it's MMENSURLSessionWrapper

Commit:
Document `crashEventReporting`
Stack frame redaction

alfwatt commented on MapboxMobileEvents/MMEEvent.m:
```diff
@@ -3,305 +3,336 @@
+        NSUInteger stackHeight = errorException.callStackSymbols.count;
+        NSUInteger index = 0;
+        while (index < stackHeight) {
+            NSString *stackFrame = [NSString stringWithFormat:@"%@ %@",
```
Added redaction for symbols which aren't Mapbox or some common system libraries

Commit:
Fix session wrapper tests to cover the manual `invalidate` case

Commit:
`MMEEventsManager` now returns the `sharedInstance` from `init`
Tests use the private constructor `initShared`

1ec5 commented on LICENSE.md:
```diff
@@ -1,5 +1,5 @@
 
-mapbox-events-ios copyright (c) 2014-2017 Mapbox.
+mapbox-events-ios copyright (c) 2014-2019 Mapbox.
```
Happy new years! 🎉

1ec5 commented on Makefile:
```diff
@@ -23,3 +23,35 @@ pod-lint:
+	echo "</tt></body></html>" >> $(DOCS_INDEX)
+
+.PHONY: docs
+docs: $(DOCS_DIR) headerdoc docindex
```
Consider using [jazzy](https://github.com/realm/jazzy/) instead of HeaderDoc. Unlike HeaderDoc, jazzy takes advantage of Clang’s built-in Doxygen parsing, so it doesn’t get tripped up by things like C attributes and macros, and it doesn’t require a separate Markdown parser.

[Here’s an example jazzy setup](https://github.com/mapbox/MapboxDirections.swift/blob/master/scripts/document.sh) and [configuration](https://github.com/mapbox/MapboxDirections.swift/tree/master/docs) in MapboxDirections.swift that applies a Mapbox-branded theme and customizes the cover page, though you might want to adopt [this shared Mapbox theme](https://github.com/mapbox/jazzy-theme/) at the same time. [Here’s what it looks like](http://docs.mapbox.com/ios/api/directions/).

/ref mapbox/mapbox-gl-native#3203
/cc @friedbunny @captainbarbosa

1ec5 commented on MapboxMobileEvents/MMEEvent.h:
```diff
@@ -1,42 +1,172 @@
+    @note please use eventWithAttributes:error:
+*/
++ (instancetype)locationEventWithAttributes:(NSDictionary *)attributes instanceIdentifer:(NSString *)instanceIdentifer commonEventData:(MMECommonEventData *)commonEventData
+    MME_DEPRECATED_GOTO("use eventWithAttributes:error:", "-eventWithAttributes:error:");
```
When both a deprecation message and a replacement are specified, the warning within Xcode is duplicative. In the map and navigation SDKs, I’ve been specifying the empty string as the message when there’s a replacement.

1ec5 commented on MapboxMobileEvents/MMEEventLogReportViewController.m:
```diff
@@ -17,7 +17,7 @@ - (void)viewDidLoad {
     navBar.backgroundColor = [UIColor colorWithRed:(247.0f/255.0f) green:(247.0f/255.0f) blue:(247.0f/255.0f) alpha:1];
     
-    UINavigationItem *navItem = [[UINavigationItem alloc] initWithTitle:@"Profiler"];
+    UINavigationItem *navItem = [[UINavigationItem alloc] initWithTitle:NSLocalizedString(@"Profiler",nil)];
```
If there’s no intention to make the library localizable, then disable the localization diagnostic rather than putting in pro forma `NSLocalizedString()` usage.

1ec5 commented on MapboxMobileEvents/Reachability/MMEReachability.h:
```diff
@@ -66,7 +67,7 @@ typedef void (^NetworkUnreachable)(MMEReachability * reachability);
 +(instancetype)reachabilityForInternetConnection;
 +(instancetype)reachabilityWithAddress:(void *)hostAddress;
-+(instancetype)reachabilityForLocalWiFi;
++(instancetype)reachabilityForLocalWiFi MME_DEPRECATED_MSG("the resulting object isn't particularly useful");
```
Sad trombone.

1ec5 commented on readme.md:
````diff
@@ -1,27 +1,23 @@
 ```objc
-MMEventsManager *manager = [MMEventsManager.alloc 
+MMEEventsManager *manager = [MMEventsManager.sharedManager 
     initilizeWithAccessToken:@"your-mapbox-token" 
````
```suggestion
    initializeWithAccessToken:@"your-mapbox-token" 
```

1ec5 commented on readme.md:
```diff
@@ -34,14 +30,48 @@ manager.isDebugLoggingEnabled = (DEBUG ? YES : NO);
+
+The MapboxMobileEvents frameworks strives to contain all internal exceptions and errors in an effort to prevent errors from directly 
+impacting the end users of applications which use the framework. The framework will attempt to report them to our backend, 
+in a redacted form, for analisys by Mapbox.
```
```suggestion
in a redacted form, for analysis by Mapbox.
```

1ec5 commented on readme.md:
```diff
@@ -34,14 +30,48 @@ manager.isDebugLoggingEnabled = (DEBUG ? YES : NO);
+impacting the end users of applications which use the framework. The framework will attempt to report them to our backend, 
+in a redacted form, for analisys by Mapbox.
+
+Applications and frameworks which embed `MapboxMobileEvents.framework` can implemente the  `MMEEventsManagerDelegate` method:
```
```suggestion
Applications and frameworks which embed `MapboxMobileEvents.framework` can implement the  `MMEEventsManagerDelegate` method:
```

1ec5 commented on readme.md:
````diff
@@ -53,3 +83,8 @@ brew install carthage
 carthage bootstrap
 ```
+
+<style>
````
GitHub-flavored Markdown doesn’t support `<style>` tags, so it just shows up as gibberish at the bottom.

1ec5 replied:
On GitHub?

![github](https://user-images.githubusercontent.com/1231218/58742948-dc630a00-83dd-11e9-9887-a470fb6aac2b.png)

1ec5 commented on MapboxMobileEventsTests/MMEExceptionalDictionary.m:
```diff
@@ -0,0 +1,22 @@
+    return 0;
+}
+
+- (NSEnumerator *)keyEnumerator {
```
[NSDictionary documentation](https://developer.apple.com/documentation/foundation/nsdictionary?language=objc#1776559) states that `-objectForKey:` needs to be overridden as well.

alfwatt replied:
It does, provided you don't want the subclass to throw an exception when a caller attempts to call that method, which in this exceptional case, is the desired behavior.

1ec5 reviewed (commented):
Sorry for all the typo nitpicks at the bottom. I think jazzy would end up being more maintainable than HeaderDoc, but if you want to leave that as tail work, that’s understandable.

alfwatt committed:
Update readme.md

Co-Authored-By: Minh Nguyễn <mxn@1ec5.org>

alfwatt committed:
Update readme.md

Co-Authored-By: Minh Nguyễn <mxn@1ec5.org>

alfwatt committed:
Update readme.md

Co-Authored-By: Minh Nguyễn <mxn@1ec5.org>

alfwatt committed:
Apply suggestions from code review

Co-Authored-By: Minh Nguyễn <mxn@1ec5.org>

Commit:
Don't crash printing a debug event if the instance hasn't completely initialized yet.
Add test to expose the crasher and validate that the forced singleton constructor is working.

Commit:
Enumerate error codes

Commit:
Explicitly release the wrapped session, fix tests
